mypy/semanal.py

```diff
@@ -2132,10 +2136,48 @@ def analyze_try_stmt(self, s: TryStmt, visitor: NodeVisitor,
             s.finally_body.accept(visitor)
 
     def visit_with_stmt(self, s: WithStmt) -> None:
+        types = []  # type: List[Type]
+
+        if s.target_type:
+            actual_targets = [t for t in s.target if t is not None]
+            if len(actual_targets) == 0:
+                # We have a type for no targets
+                self.fail('Invalid type comment', s)
+            elif len(actual_targets) == 1:
+                # We have one target and one type
+                types = [s.target_type]
+            elif isinstance(s.target_type, TupleType):
+                # We have multiple targets and multiple types
+                if len(actual_targets) == len(s.target_type.items):
+                    types = s.target_type.items
+                else:
+                    # But it's the wrong number of items
+                    self.fail('Incompatible number of tuple items', s)
```

Comment:
These error cases are reachable when the number of targets differs from the number of elements in the top-level tuple:
```python
from typing import *

class A:
    def __enter__(self) -> Tuple[int, int]: pass
    def __exit__(self, x, y, z): pass

with A() as (a, b), A() as (c, d):  # type: int, int, int
    pass
```
(You'll have to change the error messages so they're not identical to the errors in `store_declared_types` in order to verify this.)